File: src/main/scala/dojo/ttt/BoardState.scala
package dojo.ttt

sealed trait BoardState2 {

  /**
   * @return true if current state represents a final game board (e.g win condition or draw)
   */
  def isTerminal: Boolean = true

}

/**
 * Means that there are still empty positions where to put a player token.
 */
case object Play extends BoardState2 {
  override def isTerminal = false
}

/**
 * The game board is full, but nobody won.
 */
case object Draw extends BoardState2

/**
 * Player X won.
 */
case object WinX extends BoardState2

/**
 * Player O won.
 */
case object WinO extends BoardState2

File: src/src/main/java/dojo/ttt/BoardState.java
package dojo.ttt;

public enum BoardState {

    PLAY(false),
    DRAW(true),
    WIN_X(true),
    WIN_O(true);

    /**
     * Terminal state means that the game has stopped - either draw or one of the players won.
     */
    public final boolean terminal;

    private BoardState(boolean terminal) {
        this.terminal = terminal;
    }

}